### ISSUE TITLE
Secure tunneling unable to create new tunnels after a while

### ISSUE THREAD

**Issue by cijo7** [awslabs/aws-iot-device-client#151]
**To Reproduce**

Run the client for some time.
Create a tunnel. It will work.
Then after sometime create new tunnel.
Eventually, the destination will stop connecting.
I will have to manually restart the client to create new tunnel.

**Comment by harishm18** [comment]
I am also facing the same issue.

**Comment by HarshGandhi-AWS** [comment]
Hello @harishm18 @cijo7 , Sorry to hear that you both are facing this issue. I am working on to recreate this issue to solve it. Few questions, 
1. Can you guys share the commit id of Device Client? 
2. After how much time do you see issues in creating new tunnel?
3. Are there any error logs generated? If yes, can you share that? 

Regards, 
Harsh Gandhi

**Comment by harishm18** [comment]
@HarshGandhi-AWS : No errors logging in service. When I tried 2 to 3 hours later service was able to catch the connection. But I tried after 6 hours then no logs of connection and in aws iot tunnel list shows Disconnected. If I restart the service then it will connect again.

**Comment by harishm18** [comment]
@cijo7 : Any alternate services available to connect?

**Comment by HarshGandhi-AWS** [comment]
Hello @cijo7 @harishm18 , I am still trying to recreate the issue. Can you provide more details about your device environment setup?
 - OS: [e.g. Ubuntu]
 - Version [e.g. 16]
 - Architecture: [e.g. x86-64 or x86-32]
 - Device Client commit: [e.g. [6cb73bf7d27a854fc7571c2f29a5570042c4b917](https://github.com/awslabs/aws-iot-device-client/commit/6cb73bf7d27a854fc7571c2f29a5570042c4b917)]

**Comment by cijo7** [comment]
i decided to stop using the service. Its stupidly costly at 5 usd per tunnel. I didn't look at pricing page. 

I installed this repo in raspberry pi, running Ubuntu .
Latest commit at the time. 

All the best.

**Comment by harishm18** [comment]
@cijo7 
```
2021-07-16T17:56:48.114Z [WARN]  {Config.cpp}: Key {template-name} was provided in the JSON configuration file with an empty value
2021-07-16T17:56:48.114Z [WARN]  {Config.cpp}: Key {csr-file} was provided in the JSON configuration file with an empty value
2021-07-16T17:56:48.114Z [WARN]  {Config.cpp}: Key {device-key} was provided in the JSON configuration file with an empty value
2021-07-16T17:56:48.114Z [WARN]  {Config.cpp}: Key {publish-topic} was provided in the JSON configuration file with an empty value
2021-07-16T17:56:48.114Z [WARN]  {Config.cpp}: Key {publish-file} was provided in the JSON configuration file with an empty value
2021-07-16T17:56:48.114Z [WARN]  {Config.cpp}: Key {subscribe-topic} was provided in the JSON configuration file with an empty value
2021-07-16T17:56:48.114Z [WARN]  {Config.cpp}: Key {subscribe-file} was provided in the JSON configuration file with an empty value
2021-07-16T17:56:48.114Z [INFO]  {Config.cpp}: Successfully fetched JSON config file: 
    {
      "thing-name":	"xxxx-xx",
      "logging":	{
        "level":	"DEBUG",
        "type":	"FILE",
        "file": "/var/log/aws-iot-device-client/aws-iot-device-client.log"
      },
      "jobs":	{
        "enabled":	false,
        "handler-directory": "/root/.aws-iot-device-client/jobs"
      },
      "tunneling":	{
        "enabled":	true
      },
      "device-defender":	{
        "enabled":	false,
        "interval": 300
      },
      "fleet-provisioning":	{
        "enabled":	false,
        "template-name": "",
        "template-parameters": "",
        "csr-file": "",
        "device-key": ""
      },
      "samples": {
        "pub-sub": {
          "enabled": false,
          "publish-topic": "",
          "publish-file": "",
          "subscribe-topic": "",
          "subscribe-file": ""
        }
      }
    }

2021-07-16T17:56:48.115Z [DEBUG] {Config.cpp}: Did not find a runtime configuration file, assuming Fleet Provisioning has not run for this device
2021-07-16T17:56:48.115Z [INFO]  {Main.cpp}: Now running AWS IoT Device Client version v1.2.0
2021-07-16T17:56:48.124Z [INFO]  {SharedCrtResourceManager.cpp}: SDK logging is disabled. Enable it with --enable-sdk-logging on the command line or logging::enable-sdk-logging in your configuration file
2021-07-16T17:56:48.125Z [DEBUG] {Retry.cpp}: Retryable function will retry until success
2021-07-16T17:56:48.139Z [INFO]  {SharedCrtResourceManager.cpp}: Establishing MQTT connection with client id Proview-57...
2021-07-16T17:56:48.715Z [INFO]  {SharedCrtResourceManager.cpp}: MQTT connection established with return code: 0
2021-07-16T17:56:48.715Z [INFO]  {SharedCrtResourceManager.cpp}: Shared MQTT connection is ready!
2021-07-16T17:56:48.716Z [INFO]  {Main.cpp}: Jobs is disabled
2021-07-16T17:56:48.716Z [INFO]  {Main.cpp}: Secure Tunneling is enabled
2021-07-16T17:56:48.716Z [INFO]  {Main.cpp}: Device Defender is disabled
2021-07-16T17:56:48.716Z [INFO]  {Main.cpp}: Pub Sub is disabled
2021-07-16T17:56:48.716Z [INFO]  {SecureTunneling.cpp}: Running Secure Tunneling!
2021-07-16T17:56:48.716Z [INFO]  {Main.cpp}: Client base has been notified that Secure Tunneling has started
2021-07-16T17:56:48.836Z [DEBUG] {SecureTunneling.cpp}: Subscribed to tunnel notification topic
2021-07-16T18:02:19.360Z [DEBUG] {SecureTunneling.cpp}: Received MQTT Tunnel Notification
2021-07-16T18:02:19.360Z [DEBUG] {SecureTunneling.cpp}: Region=us-west-2, Service=SSH
2021-07-16T18:02:20.127Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnConnectionComplete
2021-07-16T18:07:28.360Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnStreamStart
2021-07-16T18:07:28.361Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=41
2021-07-16T18:07:28.361Z [DEBUG] {TcpForward.cpp}: Not connected yet. Saving the data to send
2021-07-16T18:07:28.361Z [DEBUG] {TcpForward.cpp}: TcpForward::OnConnectionResult
2021-07-16T18:07:28.361Z [DEBUG] {TcpForward.cpp}: Flushing send buffer
2021-07-16T18:07:28.377Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:07:28.377Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=41
2021-07-16T18:07:28.377Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:07:28.378Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:07:28.378Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=1080
2021-07-16T18:07:28.378Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:07:29.084Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=1360
2021-07-16T18:07:29.084Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=48
2021-07-16T18:07:29.101Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:07:29.101Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=452
2021-07-16T18:07:29.101Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:07:29.791Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnStreamReset
2021-07-16T18:08:01.880Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnStreamStart
2021-07-16T18:08:01.881Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=41
2021-07-16T18:08:01.881Z [DEBUG] {TcpForward.cpp}: Not connected yet. Saving the data to send
2021-07-16T18:08:01.881Z [DEBUG] {TcpForward.cpp}: TcpForward::OnConnectionResult
2021-07-16T18:08:01.881Z [DEBUG] {TcpForward.cpp}: Flushing send buffer
2021-07-16T18:08:01.896Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:08:01.896Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=41
2021-07-16T18:08:01.896Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:08:01.897Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:08:01.897Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=1080
2021-07-16T18:08:01.898Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:08:02.592Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=1360
2021-07-16T18:08:02.592Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=48
2021-07-16T18:08:02.610Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:08:02.610Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=452
2021-07-16T18:08:02.610Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:08:05.479Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=16
2021-07-16T18:08:05.479Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=44
2021-07-16T18:08:05.521Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:08:05.521Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=44
2021-07-16T18:08:05.521Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:08:06.549Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=68
2021-07-16T18:08:06.550Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:08:06.550Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=52
2021-07-16T18:08:06.550Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:08:07.267Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=372
2021-07-16T18:08:07.269Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:08:07.269Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=52
2021-07-16T18:08:07.269Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:08:08.007Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=372
2021-07-16T18:08:08.008Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:08:08.008Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=52
2021-07-16T18:08:08.008Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:08:11.073Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=84
2021-07-16T18:08:11.087Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:08:11.087Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=28
2021-07-16T18:08:11.087Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:08:11.120Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:08:11.120Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=500
2021-07-16T18:08:11.120Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:08:12.032Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=112
2021-07-16T18:08:12.032Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:08:12.032Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=44
2021-07-16T18:08:12.032Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:08:12.755Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=452
2021-07-16T18:08:12.758Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:08:12.758Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=108
2021-07-16T18:08:12.758Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:08:12.758Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:08:12.758Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=332
2021-07-16T18:08:12.758Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:08:12.758Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:08:12.758Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=132
2021-07-16T18:08:12.758Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:08:12.877Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:08:12.877Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=116
2021-07-16T18:08:12.877Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:11:28.983Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=36
2021-07-16T18:11:28.984Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:11:28.984Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=36
2021-07-16T18:11:28.984Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:11:29.478Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=36
2021-07-16T18:11:29.478Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:11:29.478Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=36
2021-07-16T18:11:29.478Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:11:30.225Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=36
2021-07-16T18:11:30.225Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:11:30.225Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=36
2021-07-16T18:11:30.225Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:11:30.228Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:11:30.228Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=276
2021-07-16T18:11:30.228Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:11:30.229Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:11:30.229Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=116
2021-07-16T18:11:30.229Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:23:05.391Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=36
2021-07-16T18:23:05.392Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=36
2021-07-16T18:23:05.392Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:23:05.392Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=36
2021-07-16T18:23:05.392Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:23:05.392Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:23:05.392Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=36
2021-07-16T18:23:05.392Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:23:05.392Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:23:05.392Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=116
2021-07-16T18:23:05.392Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:23:05.393Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:23:05.393Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=116
2021-07-16T18:23:05.393Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:23:43.652Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=36
2021-07-16T18:23:43.652Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=36
2021-07-16T18:23:43.652Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:23:43.652Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=36
2021-07-16T18:23:43.652Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:23:43.652Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:23:43.652Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=36
2021-07-16T18:23:43.653Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:23:43.877Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=36
2021-07-16T18:23:43.877Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=36
2021-07-16T18:23:43.877Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:23:43.877Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=36
2021-07-16T18:23:43.877Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:23:43.877Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:23:43.877Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=36
2021-07-16T18:23:43.878Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:23:44.229Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=36
2021-07-16T18:23:44.229Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:23:44.229Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=36
2021-07-16T18:23:44.229Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:23:44.229Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:23:44.229Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=44
2021-07-16T18:23:44.229Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:23:44.233Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:23:44.233Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=176
2021-07-16T18:23:44.233Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:23:44.981Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=96
2021-07-16T18:23:44.981Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnStreamReset
2021-07-16T18:25:34.664Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnStreamStart
2021-07-16T18:25:34.664Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=41
2021-07-16T18:25:34.664Z [DEBUG] {TcpForward.cpp}: Not connected yet. Saving the data to send
2021-07-16T18:25:34.664Z [DEBUG] {TcpForward.cpp}: TcpForward::OnConnectionResult
2021-07-16T18:25:34.664Z [DEBUG] {TcpForward.cpp}: Flushing send buffer
2021-07-16T18:25:34.681Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:34.681Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=41
2021-07-16T18:25:34.681Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:34.682Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:34.682Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=1080
2021-07-16T18:25:34.682Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:35.355Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=1360
2021-07-16T18:25:35.355Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=48
2021-07-16T18:25:35.373Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:35.373Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=452
2021-07-16T18:25:35.373Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:36.065Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=16
2021-07-16T18:25:36.065Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=44
2021-07-16T18:25:36.109Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:36.109Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=44
2021-07-16T18:25:36.109Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:36.774Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=68
2021-07-16T18:25:36.775Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:36.775Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=52
2021-07-16T18:25:36.775Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:37.457Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=372
2021-07-16T18:25:37.458Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:37.458Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=52
2021-07-16T18:25:37.458Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:38.381Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=372
2021-07-16T18:25:38.382Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:38.382Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=52
2021-07-16T18:25:38.382Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:41.096Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=84
2021-07-16T18:25:41.110Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:41.110Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=28
2021-07-16T18:25:41.110Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:41.143Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:41.143Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=500
2021-07-16T18:25:41.143Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:41.791Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=112
2021-07-16T18:25:41.791Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:41.791Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=44
2021-07-16T18:25:41.791Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:42.511Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=452
2021-07-16T18:25:42.513Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:42.514Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=108
2021-07-16T18:25:42.514Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:42.514Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:42.514Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=400
2021-07-16T18:25:42.514Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:42.514Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:42.514Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=92
2021-07-16T18:25:42.514Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:42.633Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:42.633Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=116
2021-07-16T18:25:42.633Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:44.989Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=36
2021-07-16T18:25:44.989Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=36
2021-07-16T18:25:44.989Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:44.989Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=36
2021-07-16T18:25:44.990Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:44.990Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:44.990Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=36
2021-07-16T18:25:44.990Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:45.272Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=36
2021-07-16T18:25:45.272Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:45.272Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=36
2021-07-16T18:25:45.272Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:45.662Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=36
2021-07-16T18:25:45.662Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=36
2021-07-16T18:25:45.662Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:45.662Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=36
2021-07-16T18:25:45.663Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:45.663Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:45.663Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=36
2021-07-16T18:25:45.663Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:45.663Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:45.663Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=44
2021-07-16T18:25:45.663Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:45.666Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2021-07-16T18:25:45.666Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=176
2021-07-16T18:25:45.666Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnSendDataComplete
2021-07-16T18:25:46.422Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=96
2021-07-16T18:25:46.422Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnStreamReset
2021-07-16T21:03:14.181Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnConnectionShutdown
2021-07-16T21:03:14.181Z [DEBUG] {SecureTunneling.cpp}: SecureTunnelingFeature::OnConnectionShutdown
```

**Comment by harishm18** [comment]
I have started the tunnel at 16 th for 12 hours and then I have started new tunnel just now but not connected. and even I am unable  to see any log.

**Comment by harishm18** [comment]
@cijo7 : Issue still persist but issue is closed

**Comment by harishm18** [comment]
@HarshGandhi-AWS

**Comment by cijo7** [comment]
What is your use case?

**Comment by harishm18** [comment]
@cijo7 : Connecting the edge device(Nano) through the tunnel for troubleshooting.

**Comment by HarshGandhi-AWS** [comment]
Hey @harishm18 , looking into this issue again. Last time I was not able to recreate the issue and since it is not generating any error logs for you, it is difficult to find the root cause. But I'll run some more tests to recreate the issue and solve it. Thanks for reporting. 

- Harsh Gandhi

**Comment by harishm18** [comment]
@HarshGandhi-AWS : Can I get any tentative date? I am still opening a tunnel to check which costs 5 dollars every time.

**Comment by HarshGandhi-AWS** [comment]
Hello @harishm18 , this morning I was able to reproduce the issue. But I have observed that it is failing only after 15+ hours or after couple of days in some case. Device Client is working fine if I close and create a new tunnel within an hour or two. We will continue investigating this issue and will provide a fix as soon as possible.
Since to reproduce this issue it takes a lot of time, to test the fix will also require some time. I cannot provide you any estimate time right now but we will inform you over here once the fix is available. 

-Harsh Gandhi

**Comment by harishm18** [comment]
@HarshGandhi-AWS : Thank you for the update. I hope this will get fixed ASAP.

**Comment by selin-k** [comment]
I have a similar issue. Using a RPi4B running Ubuntu 18.04.5 LTS. I can successfully open a tunnel from the management console to the pi but even if the session expiry is set to 12 hours the destination proxy disconnects on its own within a couple of minutes. Similarly, the logs don't say much and just indicate a sudden connection shutdown out of nowhere.

**Comment by HarshGandhi-AWS** [comment]
Hello @harishm18 @cijo7 @selin-k ,

Thank you for being patient with use. It took us some time to reproduce and resolve this issue. Also testing the code change took some time. The latest code change should resolve the issue you guys were facing. 
I have made changes to retry the connection if the network is interrupted for any reason. To provide best customer experience, now Device Client should reconnect on network interruption without any need to resubscribe to the IoT feature topics. 
Please let us know if this resolves the issue or not and feel free to create new issue if required. I will keep this issue open for next 8 days and will close it if no response is received. 

Regards, 
Harsh Gandhi

**Comment by harishm18** [comment]
![Screenshot from 2021-09-13 09-19-34](https://user-images.githubusercontent.com/9034512/133021453-5220f8fd-abc1-4643-bd7c-0b6b7ad7c7f1.png)


@HarshGandhi-AWS  I have taken the new updated code and follwed the same steps for running the app. But I am facing different issue now. Please check the image for reference.

**Comment by HarshGandhi-AWS** [comment]
@harishm18 If you read the last 8 lines of the logs, you can see the log message saying `Received signal: (2)`. This happens when a user interrupts the execution of the software or executable on the device by pressing `Control+C`. This interrupt signal means to kill the stream process running on the device. So this is not a problem, it is an expected behavior. 
I would suggest starting the Device Client as a service to start or stop it using `systemctl` commands. That way you will accidental not stop the agent. 

~Harsh Gandhi

**Comment by harishm18** [comment]
I have seen the error in aws . It says duplicate device client id. I have
restarted it as service. I will try to create new tunnel tomorrow and check
whether it’s connecting or not.

**Comment by HarshGandhi-AWS** [comment]
It's been a while that we haven't heard back from you. We think the fix should solve the issue you were facing earlier and so I am closing this issue now. Please feel free to open a new issue with us if face any difficulties using Device Client and we will be more than happy to help you resolve the problem. 

Thank you, 
Harsh Gandhi
